ENHANCED_PERMISSIONS_EXAMPLE.md

```diff
@@ -0,0 +1,178 @@
+# Enhanced usePermissions Hook - Type Inference & Multi-Resource API
+
+## Overview
+
+We have successfully implemented the enhanced reviewer suggestions for the usePermissions hook. This implementation addresses all the peer review feedback:
+
+1. ✅ **API Simplification**: Removed complex helper functions, focus on core permission checking
+2. ✅ **Flexible Context**: Support various field naming conventions (userId vs user_id vs id vs sub)
+3. ✅ **Configurable Rules**: Projects can provide their own rules for complete customization
+4. ✅ **Type Inference**: Automatically derive types from permission rule configurations
+5. ✅ **Multi-Resource Support**: Check permissions across multiple resources simultaneously
+
+## Key Features
+
+### 1. Type Inference from Permission Rules
+
+Types are automatically inferred from the actual permission configuration:
+
+```typescript
+// Define your permission rules
+const myProjectRules = [
+  { role: "editor", resource: "document", actions: ["view", "edit", "publish"] },
+  { role: "viewer", resource: "document", actions: ["view"] },
+  { role: "admin", resource: "user", actions: ["view", "create", "delete"] },
+] as const
+
+// Types are automatically inferred:
+// - Roles: "editor" | "viewer" | "admin"
+// - Actions: "view" | "edit" | "publish" | "create" | "delete"
+// - Resources: "document" | "user"
+```
+
+### 2. Multi-Resource Permission Checking
+
+Check permissions across multiple resources in a single hook call:
+
+```typescript
+const { results, hasPermission, hasAnyPermission } = usePermissions({
+  userRole: "editor", // Type-safe - must match inferred roles
+  context: { userId: "user-123" },
+  rules: myProjectRules,
+  checks: [
+    { resource: "document", action: "view" },
+    { resource: "document", action: "edit" },
+    { resource: "document", action: "publish", data: { ownerId: "user-123" } },
+    { resource: "user", action: "view" },
+  ]
+})
+
+// results is an array of PermissionResult objects
+// Helper functions provide convenient access patterns
+```
+
+### 3. Flexible Context Support
+
+Supports various field naming conventions automatically:
+
+```typescript
+// All of these work with the same code:
+const auth0Context = { sub: "user-123", data: { document: "doc-456" } }
+const firebaseContext = { uid: "user-123", metadata: { document: "doc-456" } }
+const customContext = { userId: "user-123", attributes: { document: "doc-456" } }
+const legacyContext = { user_id: "user-123", context: { document: "doc-456" } }
+```
+
+### 4. Complete Example
+
+```typescript
+import { usePermissions } from './hooks/usePermissions'
+import { DEFAULT_QUEUE_RULES } from './hooks/usePermissions/permission_rules'
+
+// Example: Document Management System
+const documentRules = [
+  {
+    role: "owner",
+    resource: "document",
+    actions: ["view", "edit", "delete", "share"],
+    condition: (ctx) => ctx.data?.ownerId === ctx.userId
```

Comment:
It should be able to evaluate condition per action. For example the user might be able to view all documents, but they can only delete a document if their userid matches the created_by property on the document